### ISSUE TITLE
ngtcp2: build errors after 3288e9c

### ISSUE THREAD

**Issue by vszakats** [curl/curl#8978]
### I did this

Tried to build curl-for-win with the latest curl source snapshot with QUIC support.
It uses ngtcp2 0.5.0 (latest tagged release).

1. `NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE` is missing from 0.5.0 and needs a new ngctp2 release?
2. variable-size buffer fails with C89/C90.
3. Missing `sendmsg()`. This code targets Linux, and some platform guards may be necessary?

```
clang -I. -I../include -I"../../nghttp2/pkg/usr/local/include" -I"../../libssh2/pkg/usr/local/include" -I"../../libssh2/pkg/usr/local/win32" -I"../../nghttp3/pkg/usr/local/include" -I"../../ngtcp2/pkg/usr/local/include" -I"../../openssl-quic/pkg/usr/local/include" -I"../../zlib/pkg/usr/local" -I"../../brotli/pkg/usr/local/include" -I"../../libgsasl/pkg/usr/local/include" -I"../../libidn2/pkg/usr/local/include" -O3 -target x86_64-w64-mingw32 --sysroot /usr/local/opt/mingw-w64/toolchain-x86_64 -fno-ident -DCURL_STATICLIB -DCURL_ENABLE_MQTT -DHAVE_ATOMIC -DOPENSSL_SUPPRESS_DEPRECATED -DCURL_DISABLE_OPENSSL_AUTO_LOAD_CONFIG -DNGHTTP2_STATICLIB -DNGHTTP3_STATICLIB -DNGTCP2_STATICLIB -W -Wall -m64 -DBUILDING_LIBCURL -DCURL_WITH_MULTI_SSL -DUSE_NGHTTP2 -DUSE_LIBSSH2 -DHAVE_LIBSSH2_H -DUSE_NGHTTP3 -DUSE_NGTCP2 -DUSE_OPENSSL -DHAVE_OPENSSL_SRP -DUSE_TLS_SRP -DUSE_SCHANNEL -DHAVE_LIBZ -DHAVE_ZLIB_H -DHAVE_BROTLI -DUSE_GSASL -DUSE_LIBIDN2 -DUSE_WINDOWS_SSPI -DENABLE_IPV6 -D_WIN32_WINNT=0x0501 -DHAVE_LDAP_SSL -c vquic/quiche.c -o vquic/quiche.o
vquic/ngtcp2.c:957:19: error: use of undeclared identifier 'NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE'
  qs->pktbuflen = NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE * MAX_PKT_BURST;
                  ^
vquic/ngtcp2.c:1863:16: error: variable has incomplete type 'struct iovec'
  struct iovec msg_iov = {(void *)pkt, pktlen};
               ^
vquic/ngtcp2.c:1863:10: note: forward declaration of 'struct iovec'
  struct iovec msg_iov = {(void *)pkt, pktlen};
         ^
vquic/ngtcp2.c:1864:17: error: variable has incomplete type 'struct msghdr'
  struct msghdr msg = {0};
                ^
vquic/ngtcp2.c:1864:10: note: forward declaration of 'struct msghdr'
  struct msghdr msg = {0};
         ^
vquic/ngtcp2.c:1865:20: warning: implicit declaration of function 'CMSG_SPACE' is invalid in C99 [-Wimplicit-function-declaration]
  uint8_t msg_ctrl[CMSG_SPACE(sizeof(uint16_t))] = {0};
                   ^
vquic/ngtcp2.c:1865:20: error: variable-sized object may not be initialized
  uint8_t msg_ctrl[CMSG_SPACE(sizeof(uint16_t))] = {0};
                   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
vquic/ngtcp2.c:1893:17: warning: implicit declaration of function 'sendmsg' is invalid in C99 [-Wimplicit-function-declaration]
  while((sent = sendmsg(sockfd, &msg, 0)) == -1 && SOCKERRNO == EINTR)
                ^
2 warnings and 4 errors generated.
make: *** [vquic/ngtcp2.o] Error 1
```

This likely started with commit:
3288e9c6d36ad1b928e06ff3ec54fbb276f08e01

### I expected the following

Build completing without error.

### curl/libcurl version

5912da253b64de3cb2a1a229558077219b2c8a35

### operating system

Windows (cross-build)

/cc @tatsuhiro-t

**Comment by tatsuhiro-t** [comment]
Yes, NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE is not included in the released version.
I think it needs some #ifdef guards and modifications for windows.  Unfortunately, I do not have windows development environment now.

**Comment by vszakats** [comment]
I also don't have Windows, but I'm glad to test any patches via cross-build.

(I deleted the Windows labels as only 1 of the 3 issues is Windows-related, even that is rather a non-Linux (and non-BSD?) one, affecting platforms without a `sendmsg()`.)

**Comment by tatsuhiro-t** [comment]
I think getting back old code which uses send(2) to replace do_sendmsg() for windows.
For variable-sized array, is CMSG_SPACE not macro for some platforms?

**Comment by vszakats** [comment]
It appears it isn't in this case (LLVM/clang 13.0.1), though I have not much experience with newer C standards and their support accross tools. I think in libcurl the preferred C standard is C89, so the variable-size array is already an issue.

**Comment by tatsuhiro-t** [comment]
https://github.com/curl/curl/tree/ngtcp2-fix-sendmsg-build-issue tries to fix 2 issues: missing sendmsg and CMSG_SPACE is not constant expression.  @vszakats Could you try it out?

**Comment by vszakats** [comment]
This patch will make the code _compile_ for Windows, though probably it's not enough to make it actually work. It only hides the C99 and `CMSG_SPACE` issues and will need adjustments for other platforms without a `sendmsg(2)`.
```diff
--- ngtcp2-ori.c	2022-06-08 14:03:13.240773908 +0000
+++ ngtcp2.c	2022-06-08 14:11:27.776710547 +0000
@@ -53,6 +53,10 @@
 #include "vtls/keylog.h"
 #include "vtls/vtls.h"
 
+#ifndef NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE
+#define NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE 1452
+#endif
+
 /* The last 3 #include files should be in this order */
 #include "curl_printf.h"
 #include "curl_memory.h"
@@ -1860,11 +1864,24 @@
                            struct quicsocket *qs, const uint8_t *pkt,
                            size_t pktlen, size_t gsolen)
 {
+  ssize_t sent;
+#if defined(WIN32)
+  sent = send(sockfd, (const char *)pkt, pktlen, 0);
+  if(sent == SOCKET_ERROR)
+  {
+    /* FIXME: do more fine-grained checks?
+              https://docs.microsoft.com/windows/win32/api/winsock2/nf-winsock2-send */
+    failf(data, "send() returned %zd (errno %d)", sent, SOCKERRNO);
+    return CURLE_SEND_ERROR;
+  }
+  /* FIXME: use 'sent', use variables below?: */
+  (void)qs;  /* unused */
+  (void)gsolen;  /* unused */
+#else
   struct iovec msg_iov = {(void *)pkt, pktlen};
   struct msghdr msg = {0};
   uint8_t msg_ctrl[CMSG_SPACE(sizeof(uint16_t))] = {0};
   size_t ctrllen = 0;
-  ssize_t sent;
 #if defined(__linux__) && defined(UDP_SEGMENT)
   struct cmsghdr *cm;
 #endif
@@ -1921,6 +1938,7 @@
   else {
     assert(pktlen == (size_t)sent);
   }
+#endif
 
   *psent = pktlen;
```

**Comment by vszakats** [comment]
@tatsuhiro-t Thanks, this looks good!

It compiles cleanly with these two extra hunks (the temporary fixup for the missing def and casting a pointer signedness warning [1]):
```diff
--- ngtcp2-pat1.c	2022-06-08 14:22:19.295137658 +0000
+++ ngtcp2.c	2022-06-08 14:23:04.797406534 +0000
@@ -53,6 +53,10 @@
 #include "vtls/keylog.h"
 #include "vtls/vtls.h"
 
+#ifndef NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE
+#define NGTCP2_MAX_PMTUD_UDP_PAYLOAD_SIZE 1452
+#endif
+
 /* The last 3 #include files should be in this order */
 #include "curl_printf.h"
 #include "curl_memory.h"
@@ -1931,7 +1935,7 @@
 
   *psent = 0;
 
-  while((sent = send(sockfd, pkt, pktlen, 0)) == -1 && SOCKERRNO == EINTR)
+  while((sent = send(sockfd, (const char *)pkt, pktlen, 0)) == -1 && SOCKERRNO == EINTR)
     ;
 
   if(sent == -1) {
```

(If `CMSG_SPACE` is a constant when defined, we also have no C99 issue 👍.)

[1]:
```
vquic/ngtcp2.c:1934:30: warning: passing 'const uint8_t *' (aka 'const unsigned char *') to parameter of type 'const char *' converts between pointers to integer types where one is of the unique plain 'char' type and the other is not [-Wpointer-sign]
  while((sent = send(sockfd, pkt, pktlen, 0)) == -1 && SOCKERRNO == EINTR)
                             ^~~
/usr/local/opt/mingw-w64/toolchain-x86_64/x86_64-w64-mingw32/include/winsock2.h:1033:60: note: passing argument to parameter 'buf' here
  WINSOCK_API_LINKAGE int WSAAPI send(SOCKET s,const char *buf,int len,int flags);
                                                           ^
```

**Comment by vszakats** [comment]
Fixed in 927ede7edcb7b05b8e8bbf9ced6aed523ae594a7
